internal/controlplane/xds.go

```diff
@@ -117,3 +127,95 @@ func inlineBytes(bs []byte) *envoy_config_core_v3.DataSource {
 		},
 	}
 }
+
+func inlineBytesAsFilename(name string, bs []byte) *envoy_config_core_v3.DataSource {
+	ext := filepath.Ext(name)
+	name = fmt.Sprintf("%s-%x%s", name[:len(ext)], xxhash.Sum64(bs), ext)
+
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		cacheDir = filepath.Join(os.TempDir())
+	}
+	cacheDir = filepath.Join(cacheDir, "pomerium", "envoy", "files")
+	if err = os.MkdirAll(cacheDir, 0755); err != nil {
+		log.Error().Err(err).Msg("error creating cache directory, falling back to inline bytes")
+		return inlineBytes(bs)
+	}
+
+	fp := filepath.Join(cacheDir, name)
+	if _, err = os.Stat(fp); os.IsNotExist(err) {
+		err = ioutil.WriteFile(fp, bs, 0600)
+		if err != nil {
+			log.Error().Err(err).Msg("error writing cache file, falling back to inline bytes")
+			return inlineBytes(bs)
+		}
+	} else if err != nil {
+		log.Error().Err(err).Msg("error reading cache file, falling back to inline bytes")
+		return inlineBytes(bs)
+	}
+
+	return inlineFilename(fp)
+}
+
+func inlineFilename(name string) *envoy_config_core_v3.DataSource {
+	return &envoy_config_core_v3.DataSource{
+		Specifier: &envoy_config_core_v3.DataSource_Filename{
+			Filename: name,
+		},
+	}
+}
+
+func getPolicyName(policy *config.Policy) string {
+	return fmt.Sprintf("policy-%x", policy.Checksum())
+}
+
+func envoyTLSCertificateFromGoTLSCertificate(cert *tls.Certificate) *envoy_extensions_transport_sockets_tls_v3.TlsCertificate {
+	envoyCert := &envoy_extensions_transport_sockets_tls_v3.TlsCertificate{}
+	var chain bytes.Buffer
+	for _, cbs := range cert.Certificate {
+		_ = pem.Encode(&chain, &pem.Block{
```

Comment:
safe to throw away error here?

Reply:
yeah it only errors on the `Write` call, and that shouldn't happen with a buffer.